Configure directory where msbuild will copy CefSharp files

intrueder opened the pull request:
During build process msbuild copies CefSharp files to `$(TargetDir)`
Sometimes it is not suitable, and instead of override `$(TargetDir)` globally another property has been introduced. It's name is `$(CefSharpTargetDir)`
It is possible to override this property in `csproj` file
```xml
<PropertyGroup>
  <CefSharpTargetDir>C:\MyProjectOutDir\Cef</CefSharpTargetDir>
</PropertyGroup>
```

If this property should depend on `$(TargetDir)` don't forget to override property after importing `Microsoft.CSharp.targets` (`$(TargetDir)` is defined there)
```xml
<Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
<PropertyGroup>
  <CefSharpTargetDir>$(TargetDir)\Cef</CefSharpTargetDir>
</PropertyGroup>
```

Also, `$(TargetDir)` is used by default if `$(CefSharpTargetDir)` was not specified in `csproj` file

amaitland commented:
@intrueder Thanks for the contribution! I'm currently working on other things at the moment, will probably take at least a few weeks before this is reviewed.

amaitland commented:
@intrueder I've had a quick look over your `PR` (still needs a full review). How do you plan on documenting this new feature? Simply copying to another folder isn't enough to resolve the `unmanaged` dependencies as I'm sure your aware.

intrueder commented:
This feature should be used in conjunction with AnyCPU support ( https://github.com/cefsharp/CefSharp/issues/1714 ) code sample.
Actually AnyCPU support can be documented and `CefSharpTargetDir` property can be as a remark.

amaitland commented:
> This feature should be used in conjunction with AnyCPU support ( #1714 ) code sample.

It's not limited to `AnyCPU` support though.......

> Actually AnyCPU support can be documented

I have no plans to document `AnyCPU` support past what is already included in #1714 

> CefSharpTargetDir property can be as a remark.

Please write some documentation and I'll consider it for inclusion in #1714

intrueder commented:
Complete code sample. (I've modified original sample a bit, becasue `AppDomain.CurrentDomain.SetupInformation.ApplicationBase` does not work for me)

**.csproj**
```xml
<PropertyGroup>
  <CefSharpTargetDir>C:\MyProjectOutDir\Cef</CefSharpTargetDir>
</PropertyGroup>
```

**.cs**
```cs
public partial class App : Application
{
    public App()
    {
        //Add Custom assembly resolver
        AppDomain.CurrentDomain.AssemblyResolve += Resolver;

        //Any CefSharp references have to be in another method with NonInlining
        // attribute so the assembly rolver has time to do it's thing.
        InitializeCefSharp();
	}

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void InitializeCefSharp()
    {
        var settings = new CefSettings();
        settings.EnableInternalPdfViewerOffScreen();

        // Set BrowserSubProcessPath based on app bitness at runtime
        settings.BrowserSubprocessPath = GetArchitectureSpecificPath("CefSharp.BrowserSubprocess.exe");

        // Make sure you set performDependencyCheck false
        Cef.Initialize(settings, shutdownOnProcessExit: false, performDependencyCheck: false);
    }

	// Will attempt to load missing assembly from either x86 or x64 subdir
	// Required by CefSharp to load the unmanaged dependencies when running using AnyCPU
	private static Assembly Resolver(object sender, ResolveEventArgs args)
	{
		if (args.Name.StartsWith("CefSharp"))
		{
			string assemblyName = args.Name.Split(new[] { ',' }, 2)[0] + ".dll";
			string archSpecificPath = GetArchitectureSpecificPath(assemblyName);

			if (File.Exists(archSpecificPath))
			{
				return Assembly.LoadFile(archSpecificPath);
			}
		}

		return null;
	}

	private static string GetAssemblyDirectory()
	{
		string codeBase = Assembly.GetExecutingAssembly().CodeBase;
		UriBuilder uri = new UriBuilder(codeBase);
		string path = Uri.UnescapeDataString(uri.Path);
		return Path.GetDirectoryName(path);
	}

	private static string GetArchitectureSpecificPath(string fileName)
	{
		// subfolder is the same as $(CefSharpTargetDir) points to
		string cefSubfolder = "Cef";
		return Path.Combine(GetAssemblyDirectory(), cefSubfolder, Environment.Is64BitProcess ? "x64" : "x86", fileName);
	}
}
```

amaitland commented:
> AppDomain.CurrentDomain.SetupInformation.ApplicationBase does not work for me)

What problems are you seeing using `ApplicationBase`?

amaitland commented:
This will likely be included as part of `53.0.0`, should be merged before the first `-pre` release is built.

intrueder commented:
> What problems are you seeing using `ApplicationBase`?

It is probably a special case..
That's what I have:

1. Unmanaged MyApplication.exe located in C:\ProgramFiles\MyApp
2. Additional libs/plugins in subfolder C:\ProgramFiles\MyApp\Libs. They are loaded via LoadLibrary
3. Managed library ChromiumHost.dll located in C:\ProgramFiles\MyApp\Libs. It is the only who uses CefSharp.

and now, to avoid putting CefSharp files directly into C:\ProgramFiles\MyApp\Libs, I want to put them into subfolder.

ApplicationBase points to MyApplication.exe path, i.e. C:\ProgramFiles\MyApp.
I need a folder, that will be relative to assembly where CefSharp is used (ChromiumHost.dll -> C:\ProgramFiles\MyApp\Libs), `Assembly.GetExecutingAssembly().CodeBase` helps with this.

amaitland commented:
> Unmanaged MyApplication.exe located in C:\ProgramFiles\MyApp

If your just using straight `C++`, why not use `CEF` directly?

intrueder commented:
It is not `C++` actually :) and we are about to migrate all the old code to `.Net`
so, we can't use `CEF` directly

zcanann commented:
Out of curiosity, does that mean you guys are shifting to a pure C#/.NET solution?

amaitland commented:
> Out of curiosity, does that mean you guys are shifting to a pure C#/.NET solution?

If your referring to `CefSharp`, then the answer is no. There are already pure `.Net` implementations out there.

syqjyh commented:
i use CefSharp to create a wpf dialog to open a url ,and this dialog is embedded in excel. our product need call excelAddIn.dll and work on anyCPU platform . so i download the cef demo project from https://github.com/cefsharp/CefSharp.MinimalExample/tree/demo/anycpu .Now i encountered a problem that i just replace the url in MainWindow.xaml with my URL, and it works well, but after i finished all coding and dependency related according to the demo project , our project still can't work and throw a exception (An unhandled exception of type 'System.IO.FileNotFoundException' occurred in Unknown Module.
Additional information: Could not load file or assembly 'CefSharp, Version=63.0.0.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138' or one of its dependencies. The system cannot find the file specified.) but i have added the CefSharp and related dependency to the related directory. i don't know why. can someone help me. thanks. and the MIMICAddIn.txt is our product entrance csproj file and i need create the wpf dialog in the csproj Morningstar.Commodity.xL.Market.Data. 
[MIMICAddIn.txt](https://github.com/cefsharp/CefSharp/files/1919057/MIMICAddIn.txt)
[Morningstar.Commodity.xL.Market.Data.txt](https://github.com/cefsharp/CefSharp/files/1919058/Morningstar.Commodity.xL.Market.Data.txt)
